rlog/2024-06-24-gsub-largemsg-improvements.mdx

```diff
@@ -0,0 +1,396 @@
+and numerous improvements have been considered (or even incorporated) for the GossipSub protocol to enable efficient large-message propagation [[3](https://github.com/libp2p/rust-libp2p/pull/3666),[7](https://github.com/vacp2p/nim-libp2p/pull/911),[9](https://hackmd.io/@gRwfloEASH6NWWS_KJxFGQ/B18wdnNDh),[10](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.2.md?plain=1#L52)].
+
+## Problem Realization:
+
+Sending a message to N peers involves approximately $\lceil \log_D(N) \rceil$ transmission rounds, 
+with around $(D-1)^{X-1} \times D$ transmissions in each round, 
+where $X, D$ represent the round number and mesh size. 
+
+Transmitting to a higher number of peers (floodpublish) can theoretically reduce latency by increasing the transmissions in each round to $(D-1)^{X-1} \times (F+D)$, 
+where $F$ represents the number of peers included in floodpublish. 
+
+This arrangement works fine for relatively small/moderate message sizes. 
+However, as message sizes increase, a significant rise and fluctuations in network-wide message dissemination time are seen. 
+Interestingly, a higher $D$ or $F$ can also degrade performance in this situation. 
+
+Several aspects contribute to this behavior: 
+
+1. Ideally, a message transmission to a single peer concludes in $\tau_1 = \frac {L}{R}+P$ (ignoring any message processing time), 
+where $L, R, P$ represent message size, data rate, and link latency. 
+Therefore, the time required for sending a message on a 100Mbps link with 100ms latency 
+jumps from $\tau_1^{10KB} = 100.8ms$ for a 10KB message to $\tau_1^{1MB} = 180ms$ for a 1MB message. 
+For $D$ peers, the transmission time increases to $\tau_D^{1MB} = (80 \times D) + 100ms$, 
+triggering additional queuing delays (proportional to the transmission queue size) during each transmission round.  
+
+2. In practice, $\tau_1^{1MB}$ sometimes rises to several hundred milliseconds, 
+further exaggerating the abovementioned queuing delays.
+This rise is because TCP congestion avoidance limits the maximum in-flight bytes to approximately ${C_{wnd} \times MSS}$ in a single round trip time (RTT), 
+with $C_{wnd}$ increasing with the data transfer for each flow. 
+Consequently, sending the same message through a newly established (cold) connection takes longer. 
+The message transfer time lowers as the $C_{wnd}$ grows. 
+Therefore, performance-improvement practices such as floodpublish, frequent mesh adjustment, and lazy sending 
+typically result in longer than expected message dissemination times for large messages (due to cold connections). 
+It is also worth mentioning that some TCP variants reset their $C_{wnd}$ after different periods of inactivity.
+
+3. Theoretically, the message transmission time to D peers $(\tau_D)$ remains the same, 

```

Comment:
Could you elaborate on why the message transmission time to D peers remains unchanged whether the message is relayed sequentially or simultaneously?